### PR TITLE
Use TryWriteable in icu_pattern and add MultiNamedPlaceholder backend

### PR DESCRIPTION
**Pull request by sffc**
This PR makes icu_pattern compatible with TryWriteable and implements MultiNamedPlaceholder as a backend.

### CODE REVIEW

File: utils/pattern/src/frontend/mod.rs [hunk 1]
```diff
@@ -369,56 +367,31 @@ where
...
-        self.write_to(f)
-    }
+#[test]
+fn test_try_from_store_inference() {
+    use crate::SinglePlaceholder;
+    // Does NOT work:
+    // let _: Pattern<SinglePlaceholder, String> = Pattern::try_from_str("{0} days").unwrap();
+    // Works:
+    let _ = Pattern::<SinglePlaceholder, String>::try_from_str("{0} days").unwrap();
 }
```
Comment (sffc) [inline utils/pattern/src/frontend/mod.rs:397]: @Manishearth Rust compiler puzzle. Can you help make the line "Does NOT work" compile like the other line?

Comment (Manishearth) [reply utils/pattern/src/frontend/mod.rs:428]: Yeah this is just a limitation of the resolver.

Comment (sffc) [reply utils/pattern/src/frontend/mod.rs:428]: It worked before https://github.com/unicode-org/icu4x/pull/4766/commits/353f3647d305fbaa01d3f4006d808fd6f328c020 though :frowning_face:

Comment (Manishearth) [reply utils/pattern/src/frontend/mod.rs:428]: file upstream imo

Comment (Manishearth) [reply utils/pattern/src/frontend/mod.rs:428]: it's still in the realm of common resolver limitations but the GAT people might like to know

Comment (sffc) [reply utils/pattern/src/frontend/mod.rs]: This is fixed.

File: utils/writeable/src/lib.rs [hunk 2]
```diff
@@ -81,6 +81,14 @@ use core::fmt;
 
 pub use try_writeable::TryWriteable;
 
+/// Helper types for trait impls.
+pub mod adapters {
```
Comment (sffc) [inline utils/writeable/src/lib.rs:85]: Note @robertbastian: this is where we could put the `LossyWriteable` wrapper if needed.

File: utils/pattern/src/single.rs [hunk 3]
```diff
@@ -153,7 +165,10 @@ pub enum SinglePlaceholder {}
 impl crate::private::Sealed for SinglePlaceholder {}
 
 impl PatternBackend for SinglePlaceholder {
-    type PlaceholderKey = SinglePlaceholderKey;
+    type PlaceholderKey<'a> = SinglePlaceholderKey;
+    #[cfg(feature = "alloc")]
+    type PlaceholderKeyCow<'a> = SinglePlaceholderKey;
```
Comment (sffc) [inline utils/pattern/src/single.rs:170]: These are different types because one is allocky and requires serde whereas the other is fully no_std.

Comment (robertbastian) [reply utils/pattern/src/single.rs:170]: Yeah it'd be nice if `Cow` was in `core` without the owned half.

**Review by robertbastian — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**